File: android-catalogue/components/src/main/java/xyz/ksharma/prisma/components/checkbox/PrismaCheckbox.kt
package xyz.ksharma.prisma.components.checkbox

import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.selection.triStateToggleable
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.state.ToggleableState
import androidx.compose.ui.unit.dp
import xyz.ksharma.prisma.components.icons.PrismaIcons
import xyz.ksharma.prisma.coreui.LocalPrismaIsDark
import xyz.ksharma.prisma.tokens.PrismaRadius
import xyz.ksharma.prisma.tokens.PrismaSemanticColor
import xyz.ksharma.prisma.tokens.PrismaSemanticColors
import xyz.ksharma.prisma.tokens.PrismaSpacing
import xyz.ksharma.prisma.tokens.PrismaTypography

/**
 * Binary checkbox. Use [PrismaTriStateCheckbox] for parent-selection
 * indeterminate states.
 */
@Composable
public fun PrismaCheckbox(
    checked: Boolean,
    onCheckedChange: ((Boolean) -> Unit)?,
    modifier: Modifier = Modifier,
    label: String? = null,
    helperText: String? = null,
    enabled: Boolean = true,
    isError: Boolean = false,
) {
    PrismaTriStateCheckbox(
        state = if (checked) ToggleableState.On else ToggleableState.Off,
        onClick = if (onCheckedChange != null) {
            { onCheckedChange(!checked) }
        } else null,
        modifier = modifier,
        label = label,
        helperText = helperText,
        enabled = enabled,
        isError = isError,
    )
}

/**
 * Tri-state checkbox per checkbox.md spec — supports unchecked / checked /
 * indeterminate. The glyph is a check (PrismaIcons.Check) when On, a
 * horizontal bar when Indeterminate, and absent when Off.
 *
 * The entire row is the hit target (label + helper inclusive); 44dp min
 * touch target enforced via defaultMinSize.
 */
@Composable
public fun PrismaTriStateCheckbox(
    state: ToggleableState,
    onClick: (() -> Unit)?,
    modifier: Modifier = Modifier,
    label: String? = null,
    helperText: String? = null,
    enabled: Boolean = true,
    isError: Boolean = false,
) {
    val isDark = LocalPrismaIsDark.current
    val palette = paletteFor(state, enabled, isError)

    val rowModifier = if (onClick != null) {
        modifier.triStateToggleable(
            state = state,
            onClick = onClick,
            enabled = enabled,
            role = Role.Checkbox,
        )
    } else {
        modifier
    }

    Row(
        modifier = rowModifier
            .defaultMinSize(minHeight = 44.dp)
            .padding(vertical = PrismaSpacing.Sp1),
        verticalAlignment = Alignment.Top,
        horizontalArrangement = Arrangement.spacedBy(PrismaSpacing.Sp3),
    ) {
        // 20×20 box, padded down 2dp so it aligns with the first line cap of the label.
        Box(
            modifier = Modifier
                .padding(top = 2.dp)
                .size(20.dp)
                .clip(RoundedCornerShape(PrismaRadius.Sm))
                .background(palette.fill.resolve(isDark))
                .let { mod ->
                    val borderColor = palette.border?.resolve(isDark)
                    if (borderColor != null) {
                        mod.border(2.dp, borderColor, RoundedCornerShape(PrismaRadius.Sm))
                    } else {
                        mod
                    }
                },
            contentAlignment = Alignment.Center,
        ) {
            when (state) {
                ToggleableState.On -> Icon(
                    painter = painterResource(PrismaIcons.Check),
                    contentDescription = null,
                    tint = palette.glyph.resolve(isDark),
                    modifier = Modifier.size(14.dp),
                )
                ToggleableState.Indeterminate -> Box(
                    modifier = Modifier
                        .width(10.dp)
                        .height(2.dp)
                        .background(palette.glyph.resolve(isDark), RoundedCornerShape(1.dp)),
                )
                ToggleableState.Off -> Unit
            }
        }

        if (label != null) {
            Column(verticalArrangement = Arrangement.spacedBy(2.dp)) {
                Text(
                    text = label,
                    style = PrismaTypography.BodyMd,
                    color = if (enabled) {
                        PrismaSemanticColors.TextPrimary.resolve(isDark)
                    } else {
                        PrismaSemanticColors.TextDisabled.resolve(isDark)
                    },
                )
                if (helperText != null) {
                    Text(
                        text = helperText,
                        style = PrismaTypography.BodySm,
                        color = PrismaSemanticColors.TextTertiary.resolve(isDark),
                    )
                }
            }
        }
    }
}

// region — token mapping

private data class CheckboxPalette(
    val fill: PrismaSemanticColor,
    val border: PrismaSemanticColor?,
    val glyph: PrismaSemanticColor,
)

private val transparentSemantic = PrismaSemanticColor(light = Color.Transparent, dark = Color.Transparent)

private fun paletteFor(state: ToggleableState, enabled: Boolean, isError: Boolean): CheckboxPalette {
    if (!enabled) {
        return when (state) {
            ToggleableState.Off -> CheckboxPalette(
                fill = PrismaSemanticColors.SurfaceSunken,
                border = PrismaSemanticColors.BorderSubtle,
                glyph = PrismaSemanticColors.TextDisabled,
            )
            else -> CheckboxPalette(
                fill = PrismaSemanticColors.TextDisabled,
                border = null,
                glyph = PrismaSemanticColors.SurfaceBase,
            )
        }
    }

    if (isError && state == ToggleableState.Off) {
        return CheckboxPalette(
            fill = transparentSemantic,
            border = PrismaSemanticColors.StatusDangerDefault,
            glyph = PrismaSemanticColors.StatusDangerDefault,
        )
    }

    return when (state) {
        ToggleableState.Off -> CheckboxPalette(
            fill = transparentSemantic,
            border = PrismaSemanticColors.BorderStrong,
            glyph = PrismaSemanticColors.TextPrimary,
        )
        ToggleableState.On,
        ToggleableState.Indeterminate -> CheckboxPalette(
            fill = PrismaSemanticColors.AccentDefault,
            border = null,
            glyph = PrismaSemanticColors.TextOnAccent,
        )
    }
}

// endregion

File: ios-catalogue/Sources/Components/Checkbox/PrismaCheckbox.swift
import SwiftUI
import CoreUI

public enum PrismaCheckboxState: Equatable {
    case unchecked
    case checked
    case indeterminate
}

/// Tri-state checkbox per checkbox.md spec. Glyph is a check (Image(prisma: .check))
/// when checked, a horizontal bar when indeterminate, absent otherwise.
public struct PrismaCheckbox: View {
    @Binding private var state: PrismaCheckboxState
    private let label: String?
    private let helperText: String?
    private let enabled: Bool
    private let isError: Bool

    @Environment(\.colorScheme) private var scheme

    public init(
        state: Binding<PrismaCheckboxState>,
        label: String? = nil,
        helperText: String? = nil,
        enabled: Bool = true,
        isError: Bool = false
    ) {
        self._state = state
        self.label = label
        self.helperText = helperText
        self.enabled = enabled
        self.isError = isError
    }

    /// Convenience for binary-only checkboxes — bridges Bool to the tri-state.
    public init(
        checked: Binding<Bool>,
        label: String? = nil,
        helperText: String? = nil,
        enabled: Bool = true,
        isError: Bool = false
    ) {
        self._state = Binding(
            get: { checked.wrappedValue ? .checked : .unchecked },
            set: { newValue in checked.wrappedValue = (newValue == .checked) }
        )
        self.label = label
        self.helperText = helperText
        self.enabled = enabled
        self.isError = isError
    }

    public var body: some View {
        Button(action: toggle) {
            HStack(alignment: .top, spacing: PrismaSpacing.sp3) {
                box
                    .padding(.top, 2)
                if let label {
                    VStack(alignment: .leading, spacing: 2) {
                        Text(label)
                            .font(PrismaTypography.bodyMd.font)
                            .foregroundStyle(
                                enabled
                                    ? PrismaSemanticColors.textPrimary.themed(scheme)
                                    : PrismaSemanticColors.textDisabled.themed(scheme)
                            )
                        if let helperText {
                            Text(helperText)
                                .font(PrismaTypography.bodySm.font)
                                .foregroundStyle(PrismaSemanticColors.textTertiary.themed(scheme))
                        }
                    }
                }
                Spacer(minLength: 0)
            }
            .padding(.vertical, PrismaSpacing.sp1)
            .frame(minHeight: 44, alignment: .topLeading)
        }
        .buttonStyle(.plain)
        .disabled(!enabled)
        .accessibilityLabel(label ?? "")
        .accessibilityValue(accessibilityValue)
        .accessibilityAddTraits(.isButton)
    }

    private var box: some View {
        ZStack {
            RoundedRectangle(cornerRadius: PrismaRadius.sm)
                .fill(palette.fill.themed(scheme))
            if let border = palette.border {
                RoundedRectangle(cornerRadius: PrismaRadius.sm)
                    .strokeBorder(border.themed(scheme), lineWidth: 2)
            }
            switch state {
            case .checked:
                Image(prisma: .check)
                    .renderingMode(.template)
                    .resizable()
                    .frame(width: 14, height: 14)
                    .foregroundStyle(palette.glyph.themed(scheme))
            case .indeterminate:
                RoundedRectangle(cornerRadius: 1)
                    .fill(palette.glyph.themed(scheme))
                    .frame(width: 10, height: 2)
            case .unchecked:
                EmptyView()
            }
        }
        .frame(width: 20, height: 20)
    }

    private func toggle() {
        switch state {
        case .checked, .indeterminate: state = .unchecked
        case .unchecked: state = .checked
        }
    }

    private var accessibilityValue: String {
        switch state {
        case .checked: return "checked"
        case .indeterminate: return "mixed"
        case .unchecked: return "unchecked"
        }
    }

    // MARK: - Token mapping

    private struct Palette {
        let fill: PrismaSemanticColor
        let border: PrismaSemanticColor?
        let glyph: PrismaSemanticColor
    }

    private static let transparentSemantic = PrismaSemanticColor(light: .clear, dark: .clear)

    private var palette: Palette {
        if !enabled {
            switch state {
            case .unchecked:
                return .init(fill: PrismaSemanticColors.surfaceSunken, border: PrismaSemanticColors.borderSubtle, glyph: PrismaSemanticColors.textDisabled)
            case .checked, .indeterminate:
                return .init(fill: PrismaSemanticColors.textDisabled, border: nil, glyph: PrismaSemanticColors.surfaceBase)
            }
        }
        if isError && state == .unchecked {
            return .init(
                fill: Self.transparentSemantic,
                border: PrismaSemanticColors.statusDangerDefault,
                glyph: PrismaSemanticColors.statusDangerDefault
            )
        }
        switch state {
        case .unchecked:
            return .init(fill: Self.transparentSemantic, border: PrismaSemanticColors.borderStrong, glyph: PrismaSemanticColors.textPrimary)
        case .checked, .indeterminate:
            return .init(fill: PrismaSemanticColors.accentDefault, border: nil, glyph: PrismaSemanticColors.textOnAccent)
        }
    }
}
